source/credentials_provider_process.c

```diff
@@ -0,0 +1,374 @@
+        goto done;
+    }
+
+    cJSON *creds_expiration =
+        cJSON_GetObjectItemCaseSensitive(document_root, aws_string_c_str(s_creds_expiration_name));
+    if (!cJSON_IsString(creds_expiration) || (creds_expiration->valuestring == NULL)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER, "Failed to parse Expiration from process output Json document.");
+        goto done;
+    }
+
+    /*
+     * Build the credentials
+     */
+    struct aws_byte_cursor access_key_id_cursor = aws_byte_cursor_from_c_str(access_key_id->valuestring);
+    struct aws_byte_cursor secret_access_key_cursor = aws_byte_cursor_from_c_str(secret_access_key->valuestring);
+    struct aws_byte_cursor session_token_cursor = aws_byte_cursor_from_c_str(session_token->valuestring);
+    struct aws_byte_cursor creds_expiration_cursor = aws_byte_cursor_from_c_str(creds_expiration->valuestring);
+
+    if (access_key_id_cursor.len == 0 || secret_access_key_cursor.len == 0 || session_token_cursor.len == 0) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Process credentials provider received unexpected output,"
+            " either access key, secret key or token is empty.")
+        goto done;
+    }
+
+    credentials = aws_credentials_new_from_cursors(
+        allocator, &access_key_id_cursor, &secret_access_key_cursor, &session_token_cursor);
+
+    if (credentials == NULL) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Process credentials provider failed to allocate memory for credentials.");
+        parse_error = false;
+        goto done;
+    }
+
+    if (creds_expiration_cursor.len != 0) {
+        struct aws_date_time expiration;
+        if (aws_date_time_init_from_str_cursor(&expiration, &creds_expiration_cursor, AWS_DATE_FORMAT_ISO_8601) ==
+            AWS_OP_ERR) {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "Expiration in command output Json data is not a valid ISO_8601 date string.");
+            aws_credentials_destroy(credentials);
+            credentials = NULL;
+            goto done;
+        }
+        credentials->expiration_timepoint_seconds = (uint64_t)aws_date_time_as_epoch_secs(&expiration);
+    }
+    success = true;
+done:
+    if (!success && parse_error) {
+        aws_raise_error(AWS_AUTH_PROVIDER_PARSER_UNEXPECTED_RESPONSE);
+    }
+
+    if (document_root != NULL) {
+        cJSON_Delete(document_root);
+    }
+
+    return credentials;
+}
+
+static int s_get_credentials_from_process(
+    struct aws_credentials_provider *provider,
+    aws_on_get_credentials_callback_fn callback,
+    void *user_data) {
+
+    struct aws_credentials_provider_process_impl *impl = provider->impl;
+    struct aws_credentials *credentials = NULL;
+    struct aws_run_command_options options = {.command = aws_string_c_str(impl->command)};
```

Comment:
put a comma after the closing paranthese to get the nice multi-line formatting

come to think of it, the command member should have been a byte cursor

Reply:
the command member of options? or impl? If it's impl, then I disagree, we need to make sure the memory store command won't disappear unexpectedly.